python/ray/llm/_internal/batch/stages/vllm_engine_stage.py

```diff
@@ -183,6 +191,27 @@ def _prepare_llm_request(self, row: Dict[str, Any]) -> vLLMEngineRequest:
         else:
             image = []
 
+        # The request is for LoRA.
+        lora_request = None
+        if "model" in row and row["model"] != self.model:
+            if self.vllm_use_v1:
+                raise ValueError("We do not support LoRA with vLLM v1")
+
+            lora_name = row["model"]
+            if lora_name not in self.lora_name_to_request:
+                # Load a new LoRA adapter if it is not loaded yet.
+                with self.lora_lock:
+                    # Make sure no other thread has loaded the same LoRA adapter.
+                    if lora_name not in self.lora_name_to_request:
+                        lora_request = vllm.lora.request.LoRARequest(
```

Comment:
big question. Who actually has downloaded the lora adaptor weights? Are those stored in s3? in local? That part I don't understand with the lora support.

Reply:
for base model, vllm engine does the downloading which comes from huggingface usually. What about lora here?

Reply:
You can download LoRA from HF as well. For example: https://huggingface.co/EdBergJr/Llama32_Baha_3. If the LoRA weights are not available in HF, users have to download it first and specify a local path (in a shared file system). Later we could support loading LoRA weights from S3 directly in vLLM.

Reply:
We should document these in the follow up. I think there is so many gotchas here. I know because we have solved all of these issues with the online serving and it's non-trivial. Especially around loading stuff from s3.

In online we have to have a lora-aware router (which we already do through serve multiplex router). Here in ray data, I don't know how that looks like. Maybe we should do a thread safe lora download per node here. or something similar. It needs a bit more thinking (non-blocking to this PR obv)